bpython/autocomplete.py

```diff
@@ -543,13 +546,31 @@ def get_completer(completers, cursor_offset, line, **kwargs):
         complete_magic_methods is a bool of whether we ought to complete
             double underscore methods like __len__ in method signatures
     """
-
     for completer in completers:
         matches = completer.matches(cursor_offset, line, **kwargs)
         if matches is not None:
-            return sorted(matches), (completer if matches else None)
+            if len(matches) > 0:
+                matches = sort_by_underscore(matches)
+            return matches, (completer if matches else None)
     return [], None
 
+def sort_by_underscore(matches):
+    """Sort single underscore attributes before double underscore ones.
+    """
+    matches = sorted(matches)
+    if len(matches) == 0 or len(matches) == 1:
+        return matches
+    one_underscore = None
+    for i in xrange(1, len(matches)+1):
```

Comment:
Ah ok so these won't actually work, but if there is some nice way to do
this in several steps without the for loop it could be nice. Either way,
the only real blocker right now is getting it passing the tests and getting
prints/comments out.

On Thu, Sep 10, 2015 at 12:20 PM, Wes E. Vial <notifications@github.com>
wrote:

> In bpython/autocomplete.py
> <https://github.com/bpython/bpython/pull/561#discussion_r39180057>:
>
> >      return [], None
> >
> > +def sort_by_underscore(matches):
> > +    """Sort single underscore attributes before double underscore ones.
> > +    """
> > +    matches = sorted(matches)
> > +    if len(matches) == 0 or len(matches) == 1:
> > +        return matches
> > +    one_underscore = None
> > +    for i in xrange(1, len(matches)+1):
>
> Also, [m for m in matches if m.startswith('_')] still selects the double
> underscore attributes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bpython/bpython/pull/561/files#r39180057>.
>